core/state_connector_test.go

```diff
@@ -0,0 +1,153 @@
+	accessList types.AccessList
+}
+
+func (m *mockMessage) From() common.Address         { return m.from }
+func (m *mockMessage) To() *common.Address          { return &m.to }
+func (m *mockMessage) GasPrice() *big.Int           { return m.gasPrice }
+func (m *mockMessage) GasFeeCap() *big.Int          { return m.gasFeeCap }
+func (m *mockMessage) GasTipCap() *big.Int          { return m.gasTipCap }
+func (m *mockMessage) Value() *big.Int              { return m.value }
+func (m *mockMessage) Gas() uint64                  { return m.gas }
+func (m *mockMessage) Nonce() uint64                { return 0 }
+func (m *mockMessage) Data() []byte                 { return m.data }
+func (m *mockMessage) AccessList() types.AccessList { return m.accessList }
+func (m *mockMessage) IsFake() bool                 { return true }
+
+type mockStateConnectorCaller struct {
+	context  vm.BlockContext
+	testCase string
+	msg      *mockMessage
+}
+
+func (m *mockStateConnectorCaller) Call(caller vm.ContractRef, addr common.Address, input []byte, gas uint64,
+	value *big.Int) (ret []byte, leftOverGas uint64, err error) {
+	switch {
+	case m.testCase == "happy_path" || m.testCase == "finalized_data_error":
+		voter1 := common.BytesToHash([]byte("voter1"))
+		voter2 := common.BytesToHash([]byte("voter2"))
+		merkleRootHash := []byte("some_hash")
+
+		if caller == vm.AccountRef(m.msg.From()) {
+			if addr == common.HexToAddress(GetPrioritisedFTSOContract(nil)) {
+				return []byte("voter"), 0, nil
+			}
+			if addr == common.BytesToAddress([]byte("voter")) {
+				return append([]byte{}, append(voter1.Bytes(), voter2.Bytes()...)...), 0, nil
+			}
+		}
+		if caller == vm.AccountRef(common.BytesToAddress(voter1.Bytes())) ||
+			caller == vm.AccountRef(common.BytesToAddress(voter2.Bytes())) {
+			return merkleRootHash, 0, nil
+		}
+		if caller == vm.AccountRef(stateConnectorCoinbaseSignalAddr(nil, nil)) {
+			if m.context.Coinbase != stateConnectorCoinbaseSignalAddr(nil, nil) {
+				return nil, 0, fmt.Errorf("invalid coinbase address: expected %v, got %v", stateConnectorCoinbaseSignalAddr(nil, nil), m.context.Coinbase)
+			}
+			if m.testCase == "happy_path" {
+				return nil, 0, nil
+			}
+			return nil, 0, errors.New("finalization error")
+		}
+	case m.testCase == "default_attestors_error_1":
+		return nil, 0, errors.New("voter whitelister error")
+	case m.testCase == "default_attestors_error_2":
+		switch addr {
+		case common.HexToAddress(GetPrioritisedFTSOContract(nil)):
+			return []byte("voter"), 0, nil
+		case common.BytesToAddress([]byte("voter")):
+			return nil, 0, errors.New("list price provider error")
+		}
+	}
+	return nil, 0, errors.New("undefined test case")
+}
+
+func (m *mockStateConnectorCaller) WithBlockContext(bc vm.BlockContext) { m.context = bc }
+
+func (m *mockStateConnectorCaller) BlockContext() vm.BlockContext { return m.context }
+
+func TestStateTransition_FinalisePreviousRound(t *testing.T) {
+	t.Run("happy_path", func(t *testing.T) {
```

Comment:
We usually describe the subtest with sentences instead of snake-cased names. Here we would usually call this `nominal case`, and other cases `handles voter whitelister error` and so on. It's more descriptive than `default_attestors_error_1`